lib/sycamore/sycamore/utils/xycut.py

```diff
@@ -0,0 +1,271 @@
+            next = self.elists[1]  # noqa: F841
+            return isinstance(elem, Element)
+        except (IndexError, TypeError):
+            return False
+
+    def to_str_impl(self, pfx: str, sio: StringIO) -> None:
+        for idx, elist in enumerate(self.elists):
+            for i, elem in enumerate(elist):
+                s = elem_to_str(elem)
+                sio.write(f"{pfx}{i}: {s}\n")
+
+    def to_text_impl(self, sio: StringIO) -> None:
+        for elist in self.elists:
+            for elem in elist:
+                if tr := elem.text_representation:
+                    sio.write(tr)
+                    sio.write("\n\n")
+
+    def to_elems_impl(self, elems: ElemList) -> None:
+        for elist in self.elists:
+            elems.extend(elist)
+
+
+###############################################################################
+
+
+def get_bbox(elem: Element) -> tuple[float, float, float, float]:
+    if bbox := elem.data.get("bbox"):
+        return bbox
+    return (1.0, 1.0, 1.0, 1.0)
+
+
+def bbox_to_str(bbox: tuple[float, float, float, float]) -> str:
+    a, b, c, d = [int(1000 * x) for x in bbox]
+    return f"({a:03d},{b:03d})({c:03d},{d:03d})"
+
+
+def elem_to_str(elem: Element) -> str:
+    s = bbox_to_str(get_bbox(elem))
+    if tr := elem.text_representation:
+        nonl = tr.replace("\n", " ")[:32]
+        s += f"[{nonl}]"
+    return s
+
+
+def make_begin_end(elems: ElemList, axis: int) -> BeginEndList:
+    """Returns array of (coord, isopen, elem)"""
+    ary: BeginEndList = []
+    for elem in elems:
+        bbox = get_bbox(elem)
+        aa = bbox[axis]
+        bb = bbox[axis + 2]
+        if bb < aa:
+            aa, bb = bb, aa
+        ary.append((aa, OPEN, elem))
+        ary.append((bb, CLOSE, elem))
+    ary.sort()
+    return ary
+
+
+def gen_overlap(ary: BeginEndList) -> Generator[tuple, None, None]:
+    """Yields tuple (coord, isopen, elem, count, width)"""
+    count = 0
+    cur = ary[0]
+    n = len(ary)
+    for ii in range(n):
+        width = -1.0
+        if cur[1] == OPEN:
+            count += 1
+            next = ary[ii + 1]
+        else:
+            count -= 1
+            if (ii + 1) < n:
+                next = ary[ii + 1]
+                if count == 0:
+                    width = next[0] - cur[0]
+        yield (*cur, count, width)
+        cur = next
+
+
+def widest_cut(order: BeginEndList) -> tuple[float, Element]:
+    rv = (-1.0, order[0][2])
+    if len(order) < 3:  # order is twice as long as elems
+        return rv
+    for _, _, elem, count, width in gen_overlap(order):
+        if count == 0:
+            rv = max(rv, (width, elem))
+    return rv
+
+
+def choose_axis(elems: ElemList) -> Optional[tuple[BeginEndList, Element]]:
+    xorder = make_begin_end(elems, XAXIS)
+    yorder = make_begin_end(elems, YAXIS)
+    xw, xe = widest_cut(xorder)
+    yw, ye = widest_cut(yorder)
+    if max(xw, yw) < 0.0:
+        return None
+    if xw < yw:
```

Comment:
remove debugging or change it to:
```
if False:
    ... debugging stuff ...
```
to make it easier to turn on and to have typechecking check it.